### PR TITLE
#248 serializable callback

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
```
import typing as t
from taipy.gui import Gui, State, get_context_id, invoke_state_callback

val = 1

aGui = Gui("""
#Test CB
<|run CB|button|id=my_but|on_action=on_bt_action|>
<|{val}|>
""")

def user_callback(state: State, scenario, job):
    print(f"In user_callback(state, scenario: {scenario}, job: {job})")
    state.val = state.val * 10

def on_serializable_callback(scenario, job, context_id: str):
    invoke_state_callback(aGui, context_id, user_callback, (scenario, job))

def on_bt_action(state: State, id: str, action: str, payload: t.Any):
    print(f"on_bt_action(state: State, id: {id}, action: {action}, payload: {payload})")

def on_init(state: State):
    scenario.subscribe(on_serializable_callback, {"context_id": get_context_id(state)})

aGui.run(use_reloader=False)

```

### CODE REVIEW

File: taipy/gui/gui.py [hunk 1]
```diff
@@ -679,6 +679,15 @@ def _call_function_with_state(self, user_function: t.Callable, args: t.List[t.An
             args = args[:arg_count]
         return user_function(*args)
 
+    def _call_user_callback(self, context_id: t.Optional[str], user_callback: t.Callable, args: t.List[t.Any]):
+        self._set_client_id(context_id)
+        try:
+            self._call_function_with_state(user_callback, args)
+        except Exception as e:
+            warnings.warn(f"'invoke_state_callback() cannot execute {user_callback.__name__}\n{e}")
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]: Inconsistent message, compared to what we use for User Callback.
It's not that we cannot execute... it's just that it raised an exception.

File: taipy/gui/gui_actions.py [hunk 2]
```diff
@@ -133,3 +133,17 @@ def get_context_id(state: State) -> t.Optional[str]:
     if state and isinstance(state._gui, Gui):
         return state._gui._get_client_id()
     return None
+
+def invoke_state_callback(gui: Gui, context_id: str, user_callback: t.Callable, args: t.Union[t.Tuple, t.List]):
+    """Invoke a user defined function with state (^State) as first parameter and args other parameters
```
Comment (FabienLelaquais) [inline taipy/gui/gui_actions.py]: ->
Invoke a user callback with context.

File: taipy/gui/gui_actions.py [hunk 3]
```diff
@@ -133,3 +133,17 @@ def get_context_id(state: State) -> t.Optional[str]:
...
+
+    Arguments:
+        gui (Gui^): The current gui instance.
+        context_id: The context id as returned by ^get_context_id.
+        user_callback (Callable[[State, ...], None): A user defined function that takes parameters of which the first one should be a state (^State).
```
Comment (FabienLelaquais) [inline taipy/gui/gui_actions.py]: ->
The user-defined function that is invoked.<br/>
The first parameter of this function must be a `State^`.

File: taipy/gui/gui_actions.py [hunk 4]
```diff
@@ -133,3 +133,17 @@ def get_context_id(state: State) -> t.Optional[str]:
     if state and isinstance(state._gui, Gui):
         return state._gui._get_client_id()
     return None
+
+def invoke_state_callback(gui: Gui, context_id: str, user_callback: t.Callable, args: t.Union[t.Tuple, t.List]):
+    """Invoke a user defined function with state (^State) as first parameter and args other parameters
+
+    Arguments:
+        gui (Gui^): The current gui instance.
+        context_id: The context id as returned by ^get_context_id.
```
Comment (FabienLelaquais) [inline taipy/gui/gui_actions.py]: ... as returned by `get_context_id()^`.

File: taipy/gui/gui_actions.py [hunk 5]
```diff
@@ -133,3 +133,17 @@ def get_context_id(state: State) -> t.Optional[str]:
...
+        user_callback (Callable[[State, ...], None): A user defined function that takes parameters of which the first one should be a state (^State).
+        args: A list/tuple of arguments.
```
Comment (FabienLelaquais) [inline taipy/gui/gui_actions.py]: args: The remaining arguments, as a List or a Tuple.

**Review by FabienLelaquais — APPROVED** [review]
Minor doc changes.
Great!

**Review by dinhlongviolin1 — APPROVED** [review]
Nicely done!

**Pull request merged**